x/exchange/orders_test.go

```diff
@@ -0,0 +1,2786 @@
+			var actual byte
+			testFunc := func() {
+				actual = tc.order.GetOrderTypeByte()
+			}
+			assertions.RequirePanicEquals(t, testFunc, tc.expPanic, "GetOrderTypeByte()")
+			assert.Equal(t, tc.expected, actual, "GetOrderTypeByte() result")
+		})
+	}
+}
+
+func TestOrder_GetHoldAmount(t *testing.T) {
+	tests := []struct {
+		name     string
+		order    *Order
+		expected sdk.Coins
+		expPanic string
+	}{
+		{
+			name: "AskOrder",
+			order: NewOrder(1).WithAsk(&AskOrder{
+				Assets:                  sdk.NewInt64Coin("acorns", 85),
+				SellerSettlementFlatFee: &sdk.Coin{Denom: "bananas", Amount: sdkmath.NewInt(12)},
+			}),
+			expected: sdk.NewCoins(sdk.NewInt64Coin("acorns", 85), sdk.NewInt64Coin("bananas", 12)),
+		},
+		{
+			name: "BidOrder",
+			order: NewOrder(2).WithBid(&BidOrder{
+				Price:               sdk.NewInt64Coin("acorns", 85),
+				BuyerSettlementFees: sdk.NewCoins(sdk.NewInt64Coin("bananas", 4), sdk.NewInt64Coin("cucumber", 7)),
+			}),
+			expected: sdk.NewCoins(
+				sdk.NewInt64Coin("acorns", 85),
+				sdk.NewInt64Coin("bananas", 4),
+				sdk.NewInt64Coin("cucumber", 7),
+			),
+		},
+		{
+			name:     "nil inside order",
+			order:    NewOrder(3),
+			expPanic: nilSubTypeErr(3),
+		},
+		{
+			name:     "unknown order type",
+			order:    newUnknownOrder(4),
+			expPanic: unknownSubTypeErr(4),
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var actual sdk.Coins
+			testFunc := func() {
+				actual = tc.order.GetHoldAmount()
+			}
+			assertions.RequirePanicEquals(t, testFunc, tc.expPanic, "GetHoldAmount()")
+			assert.Equal(t, tc.expected.String(), actual.String(), "GetHoldAmount() result")
+		})
+	}
+}
+
+func TestOrder_Validate(t *testing.T) {
+	// Annoyingly, sdkmath.Int{} (the zero value) causes panic whenever you
+	// try to do anything with it. So we have to give it something.
+	zeroCoin := sdk.Coin{Amount: sdkmath.ZeroInt()}
+	tests := []struct {
+		name  string
+		Order *Order
+		exp   []string
+	}{
+		{
+			name:  "order id is zero",
+			Order: NewOrder(0),
+			exp:   []string{"invalid order id", "must not be zero"},
+		},
+		{
+			name:  "nil sub-order",
+			Order: NewOrder(1),
+			exp:   []string{nilSubTypeErr(1)},
+		},
+		{
+			name:  "unknown sub-order type",
+			Order: newUnknownOrder(3),
+			exp:   []string{unknownSubTypeErr(3)},
+		},
+		{
+			name:  "ask order error",
+			Order: NewOrder(1).WithAsk(&AskOrder{MarketId: 0, Price: zeroCoin}),
+			exp:   []string{"invalid market id", "must not be zero"},
+		},
+		{
+			name:  "bid order error",
+			Order: NewOrder(1).WithBid(&BidOrder{MarketId: 0, Price: zeroCoin}),
+			exp:   []string{"invalid market id", "must not be zero"},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			err := tc.Order.Validate()
+			assertions.AssertErrorContents(t, err, tc.exp, "Validate() error")
+		})
+	}
+}
+
+func TestOrder_Split(t *testing.T) {
+	coin := func(amount int64, denom string) sdk.Coin {
```

Comment:
Is this the same as sdk.NewInt64Coin?

Reply:
See https://github.com/provenance-io/provenance/pull/1661#discussion_r1362633596